packages/kbn-object-versioning/lib/object_transform.ts

```diff
@@ -0,0 +1,195 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { ObjectMigrationDefinition, ObjectTransform, ObjectTransforms, Version } from './types';
+import { validateObj, validateVersion } from './utils';
+
+/**
+ * Extract versions metadata from an object migration definition
+ *
+ * @param migrationDefinition The object migration definition
+ * @returns Metadata about the versions (list of available, last supported, latest)
+ */
+const getVersionsMeta = (migrationDefinition: ObjectMigrationDefinition) => {
+  const versions = Object.keys(migrationDefinition)
+    .map((version) => {
+      const { result, value } = validateVersion(version);
+      if (!result) {
+        throw new Error(`Invalid version number [${version}].`);
+      }
+      return value as number;
+    })
+    .sort((a, b) => a - b);
+
+  const latestVersion = versions[versions.length - 1];
+  const lastSupportedVersion = versions[0];
+
+  return {
+    versions,
+    lastSupportedVersion,
+    latestVersion,
+  };
+};
+
+/**
+ * Get a list of pure functions to transform an object from one version
+ * to another. Either "up" or "down" according if the "to" is > than the "from"
+ *
+ * @param from The version to start from
+ * @param to The version to end to
+ * @param migrationDefinition The object migration definition
+ * @returns An array of transform functions
+ */
+const getTransformFns = (
+  from: Version,
+  to: Version,
+  migrationDefinition: ObjectMigrationDefinition
+): ObjectTransform[] => {
+  const fns: ObjectTransform[] = [];
+
+  let i = from;
+  let fn: ObjectTransform | undefined;
+  if (to > from) {
+    while (i <= to) {
+      fn = migrationDefinition[i].up;
+      if (fn) {
+        fns.push(fn);
+      }
+      i++;
+    }
+  } else if (to < from) {
+    while (i >= to) {
+      fn = migrationDefinition[i].down;
+      if (fn) {
+        fns.push(fn);
+      }
+      i--;
+    }
+  }
+
+  return fns;
+};
+
+/**
+ * Initiate a transform for a specific request version. After we initiate the transforms
+ * for a specific version we can then pass different `ObjectMigrationDefinition` to the provided
+ * handler to start up/down transforming different object based on this request version.
+ *
+ * @example
+ *
+ * ```ts
+ * const transforms = initTransform(2); // start from version "2"
+ * const fooTransforms = transforms(fooMigrationDefinition);
+ * const barTransforms = transforms(barMigrationDefinition);
+ *
+ * // Up transform the objects to the latest, starting from version "2"
+ * const { value: fooOnLatest } = foo.up();
+ * const { value: barOnLatest } = bar.up();
+ * ```
+ *
+ * @param requestVersion The starting version before up/down transforming
+ * @returns A handler to pass an object migration definition
+ */
+export const initTransform =
+  (requestVersion: Version) =>
+  (migrationDefinition: ObjectMigrationDefinition): ObjectTransforms => {
```

Comment:
Just curious, why is this function curried?

Reply:
I thought it would be nice to initiate a transform for a specific version as that's what our use case is (the browser is on a specific version and every objects must work for that specific version) and then create different transforms for different object out of it. A bit of functional programming 😊